Голубев Дмитрий

zokeer opened the pull request.

zokeer committed:
first commit

Sergey-Ubogov commented:
тащи димас

zokeer commented:
Следишь за мной?)

honest-hrundel commented:
:green_apple: Пройдено тестов 15 из 15

melanholy commented:
все следят за твоими успехами

onufrienko commented:
Привет! :) Я посмотрю твой пр сегодня вечером либо завтра, не теряй меня, если что.

zokeer commented:
Да как тебе удобно будет

onufrienko commented on lego.js:
```diff
@@ -1,70 +1,97 @@
- */
+exports.isStar = false;
+var Priors = {
+    'filterIn': 5,
```
В данном случае нет никакой необходимости оборачивать поля в кавычки

zokeer replied:
Я понимаю, что в JS так, но вообще, если обернуть, то будет нагляднее.

onufrienko replied:
Не знаю, как насчет наглядности, но не надо так :)

onufrienko commented on lego.js:
```diff
@@ -1,70 +1,97 @@
+        funcList.push(arguments[i]);
+    }
+    funcList.sort(function (one, another) {
+        return Priors[another.name] - Priors[one.name];
```
👍

onufrienko commented on lego.js:
```diff
@@ -1,70 +1,97 @@
-    return collection;
+    var friends = JSON.parse(JSON.stringify(collection));
+    var funcList = [];
+    for (var i = 1; i < arguments.length; i++) {
```
Можно для чуточку более приятного восприятия добавить пустую строку перед блоком for.
А ещё можно использовать при желании forEach :)

onufrienko commented on lego.js:
```diff
@@ -1,70 +1,97 @@
+    var friends = JSON.parse(JSON.stringify(collection));
+    var funcList = [];
+    for (var i = 1; i < arguments.length; i++) {
+        funcList.push(arguments[i]);
```
А можно и map попробовать

onufrienko commented on lego.js:
```diff
@@ -1,70 +1,97 @@
+
 exports.select = function () {
-    return;
+    var selectedFields = Array.from(arguments);
```
ES6? :)

zokeer replied:
Миру не угоден `Array.from`?))

onufrienko commented on lego.js:
```diff
@@ -1,70 +1,97 @@
+
+    return function select(collection) {
+        collection.forEach(function (elem) {
+            var properties = Object.keys(elem);
```
Ты используешь переменную properties единожды, может, в ней нет смысла?
Или ты это сделал для более удобного восприятия?

zokeer replied:
Именно, цепочка вызовов keys и проход по ним показались мне достаточно непонятными для чтения.

onufrienko replied:
Хорошо, давай оставим так

onufrienko commented on lego.js:
```diff
@@ -1,70 +1,97 @@
+            });
+        });
+
+        return collection;
```
Можно же и  return collection.forEach(function (elem)) ...

onufrienko commented on lego.js:
```diff
@@ -1,70 +1,97 @@
+            return false;
+        });
+
+        return collection;
```
Можно попробовать сразу возвращать collection.filter()

onufrienko commented on lego.js:
```diff
@@ -1,70 +1,97 @@
+    return function sortBy(collection) {
+        collection.sort(function (one, another) {
+            if (one.hasOwnProperty(property) && another.hasOwnProperty(property)) {
+                return one[property] > another[property] ? 1 : -1;
```
Чтобы избавиться от collection.reverse() ниже, можно просто здесь умножать на 1 или -1 в зависимости от порядка

zokeer replied:
Хорошо, но мне кажется, что это какая-то не очевидная константа. У меня медленно, но зато наглядно) Но я переделаю)

onufrienko commented:
🍅

zokeer committed:
corrected mistakes

honest-hrundel commented:
:green_apple: Пройдено тестов 15 из 15

onufrienko commented on lego.js:
```diff
@@ -2,19 +2,16 @@
-    for (var i = 1; i < arguments.length; i++) {
-        funcList.push(arguments[i]);
-    }
+    var funcList = Array.from(arguments).slice(1);
```
плюшки ES6, конечно, крутые, но правильно ли я помню, что вам пока не дали добро ими пользоваться в домашках? :)

zokeer replied:
нельзя использовать библиотеки, решающие большую часть задания, про ES6 ничего не говорили, да и эта штука не такая уж и кардинальная, легко записать одним циклом, а так выглядит приличнее. Думаешь, надо поменять на forEach?

onufrienko replied:
таки нашла :)
![image](https://cloud.githubusercontent.com/assets/14929766/20032604/aa90c91c-a39e-11e6-888e-0f28c6a80328.png)

onufrienko replied:
я соглашусь, что так выглядит гораздо лучше. И ништяки es6 сама ценю и люблю, но давай пока обойдемся старым добрым циклом или же чего-либо в духе
var args = Array.prototype.slice.call(arguments, 1);

zokeer replied:
Хорошо, сейчас перезалью, спасибо за поиск)

zokeer committed:
removed ES6

honest-hrundel commented:
:green_apple: Пройдено тестов 15 из 15

onufrienko commented:
:rocket:

dzlk commented on lego.js:
```diff
@@ -1,70 +1,89 @@
- * @returns {Array}
- */
+exports.isStar = false;
+var Priors = {
```
❗️ С большой буквы обычно называют функции-конструкторы. Тут это просто объект и должен быть назван с маленькой

dzlk commented on lego.js:
```diff
@@ -1,70 +1,89 @@
+    select: 3,
+    sortBy: 4,
+    format: 1,
+    limit: 2
```
почему функции не разместить в порядке убывания их значения? Сейчас страдает читаемость

dzlk commented on lego.js:
```diff
@@ -1,70 +1,89 @@
+
 exports.query = function (collection) {
-    return collection;
+    var friends = JSON.parse(JSON.stringify(collection));
```
❗️ Так клонировать объекты нельзя. JSON сильно ограниченнее формат, чем javascript, поэтому ты можешь потерять какие-нибудь свойства. Самый простой пример:

```
> JSON.parse(JSON.stringify({ a: undefined }))
{}
```

dzlk commented on lego.js:
```diff
@@ -1,70 +1,89 @@
+        friends = func(friends);
+    });
+
+    return friends;
```
❗️Функция `sort` возвращает массив, поэтому методы можно объединить в цепочку. Фукнция `forEach` здесь подходит хуже чем `reduce`, так как по сути тебе нужно выполнить каждое действие, аккумулируя результат
Хочу увидеть примерно следующее

```
returns funcList
    .sort(...)
    .reduce(...)
```

dzlk commented on lego.js:
```diff
@@ -1,70 +1,89 @@
+                    delete elem[property];
+                }
+            });
+        });
```
❗️ Свойств для выбора будет прилетать скорее всего меньше, чем полей в объекте. Поэтому лучше итерироваться по ним и заодно проверять существовать свойства на объекте. Это позволит убрать вызов `Object.keys` и сделать код чуточку читаемее

dzlk commented on lego.js:
```diff
@@ -1,70 +1,89 @@
-    console.info(property, values);
+    return function filterIn(collection) {
+        return collection.filter(function (elem) {
+            if (elem.hasOwnProperty(property)) {
```
❗️ Зачем эта проверка?

dzlk commented on lego.js:
```diff
@@ -1,70 +1,89 @@
-    return;
+    return function sortBy(collection) {
+        return collection.sort(function (one, another) {
+            if (one.hasOwnProperty(property) && another.hasOwnProperty(property)) {
```
❗️ Зачем эта проверка? Объекты одинаковой структуры – это гарантируется

dzlk commented on lego.js:
```diff
@@ -1,70 +1,89 @@
+
 exports.sortBy = function (property, order) {
-    console.info(property, order);
+    var sign = { 'asc': 1, 'desc': -1 };
```
Сложно, проще было тернарником сделать:
`sign = order === 'asc' ? 1 : -1`

dzlk commented on lego.js:
```diff
@@ -1,70 +1,89 @@
-    console.info(property, formatter);
+    return function format(collection) {
+        collection.forEach(function (elem) {
+            if (elem.hasOwnProperty(property)) {
```
👍 Молодец, что добавил проверку

dzlk commented on lego.js:
```diff
@@ -1,70 +1,89 @@
 exports.format = function (property, formatter) {
-    console.info(property, formatter);
+    return function format(collection) {
+        collection.forEach(function (elem) {
```
❗️ лучше `map`. Так как по факту ты хочешь над кажды элементом сделать операцию и вернуть новый элемент (в твоем случае тотже)

dzlk reviewed (changes requested):
Все что помечено ❗️ обязательно, остально на усмотрение

dzlk commented:
🍅

dzlk commented:
Жду смайлика зеленого яблочка, когда можно будет смотреть вновь

zokeer committed:
Fixed all mistakes

honest-hrundel commented:
:green_apple: Пройдено тестов 15 из 15

dzlk commented on lego.js:
```diff
@@ -1,70 +1,90 @@
+function copyCollection(collection) {
+    var friends = [];
+    collection.forEach(function (elem) {
+        friends.push(Object.assign({}, elem));
```
❗️ Задания мы просим выполнять в рамках спецификации ES5.1. `Object.assign` – это часть ES6. Давай без него

dzlk commented on lego.js:
```diff
@@ -1,79 +1,80 @@
+        return priors[another.name] - priors[one.name];
+    }).reduce(function (prevArr, func) {
+        return func(prevArr);
+    }, friends);
```
я обычно делаю так

``` js
return funcList
    .sort(function (one, another) {
        return priors[another.name] - priors[one.name];
    })
    .reduce(function (prevArr, func) {
        return func(prevArr);
    }, friends);
```

dzlk commented on lego.js:
```diff
@@ -1,79 +1,80 @@
+
+    return funcList.sort(function (one, another) {
+        return priors[another.name] - priors[one.name];
+    }).reduce(function (prevArr, func) {
```
❗️ не понятно почему эта переменная называется prevArr. В reduce ушла friends, а в функцию пришло collection.
При том, что prevArr слишком абстрактная, а friends слишком конкретная. Я же могу воспользоваться твоей библиотекой не для друзей?

dzlk commented on lego.js:
```diff
@@ -1,79 +1,80 @@
-            properties.forEach(function (property) {
-                if (selectedFields.indexOf(property) === -1) {
-                    delete elem[property];
+        return collection.slice().map(function (elem) {
```
❗️ Зачем slice?

dzlk commented on lego.js:
```diff
@@ -1,79 +1,80 @@
-                    delete elem[property];
+        return collection.slice().map(function (elem) {
+            var changedElem = {};
+            selectedFields.forEach(function (field) {
```
Это больше reduce. Аккумуляция свойств в объект.

dzlk commented on lego.js:
```diff
@@ -1,79 +1,80 @@
+            return elem;
         });
 
         return collection;
```
❗️ Функция map возвращает массив. Поэтому можно сразу результат возвращать

```
return collection.map()
```

dzlk reviewed (changes requested):
🍅

zokeer committed:
Corrected new mistakes

Deleted unnecessary slice();
Reworked copying;
Minor adjustments for easier reading

honest-hrundel commented:
:green_apple: Пройдено тестов 15 из 15

zokeer commented:
🍏

dzlk commented on lego.js:
```diff
@@ -9,32 +9,44 @@ var priors = {
 
+function copyElem(elem) {
+    var copied = {};
+    for (var attr in elem) {
```
❗️ Итерироваться по ключам объекта проще через Object.keys (вам об этом рассказывали).
Тут лучше сделать `return Object.keys().reduce(..., {});`

dzlk commented on lego.js:
```diff
@@ -9,32 +9,44 @@ var priors = {
     });
 
-    return friends;
+    return copiedCollection;
```
❗️ Ну map же
